ImportPlaylist: added slotCreateImportPlaylistDirect()

shivang0-0 opened the pull request:
This is in reference to #11178 

added slotCreateImportPlaylistDirect(const QString&) to handle playlist file types


This function would enable to read the playlist files independent of any qAction.
Known issue: The [following calling](https://github.com/shivang0-0/mixxx/blob/playlist-import/src/coreservices.cpp#L496) is facing issue as BasePlaylistFeature is an abstract class.

shivang0-0 committed:
ImportPlaylist: added BasePlaylistFeature::slotCreateImportPlaylistDirect(const QString&) to handle playlist file types

The following function would enable to read the playlist files independent of qAction

daschuer commented:
This is the error message: 
```
[ 51%] Building CXX object CMakeFiles/mixxx-lib.dir/src/mixxxapplication.cpp.o
/home/runner/work/mixxx/mixxx/src/coreservices.cpp: In member function ‘void mixxx::CoreServices::initialize(QApplication*)’:
Error: /home/runner/work/mixxx/mixxx/src/coreservices.cpp:496:64: error: cannot call member function ‘void BasePlaylistFeature::slotCreateImportPlaylistDirect(const QString&)’ without object
  496 |             BasePlaylistFeature::slotCreateImportPlaylistDirect(playlistFile);
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
```
The call with the class name prefix can only be used for static functions.

daschuer commented on src/coreservices.cpp:
```diff
@@ -104,6 +106,21 @@ inline QLocale inputLocale() {
+// Function to check if a given file is a playlist
+bool isPlaylist(QString fileName) {
+    // Create a list of known playlist file extensions
+    QStringList playlistExtensions;
+    playlistExtensions << "m3u"
+                       << "m3u8"
+                       << "pls";
```
This will initialize the list in a single run. 

```suggestion
    const QStringList playlistExtensions = {
            QStringLiteral("m3u"),
            QStringLiteral("m3u8"),
            QStringLiteral("pls")};
```
Not sure if we need QString(QStringLiteral("m3u")) ..

daschuer commented on src/coreservices.cpp:
```diff
@@ -528,11 +552,11 @@ void CoreServices::initializeKeyboard() {
     UserSettingsPointer pConfig = m_pSettingsManager->settings();
     if (toggle) {
-        //qDebug() << "Enable keyboard shortcuts/mappings";
+        // qDebug() << "Enable keyboard shortcuts/mappings";
```
Don't reformat unrelated code. This makes your PR unnecessary big.

ronso0 replied:
I think this is precommit, and I suggest we suppress this extra space in case for often commented function like qDebug, qWarning etc.

daschuer replied:
Pre-commit works only on changed lines. It is probably the editor that does it. It needs to be configured not to reformat the whole file on safe.

shivang0-0 replied:
I think it's the code editor's re-formatting coming to play here.

daschuer commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -536,6 +536,27 @@ void BasePlaylistFeature::slotCreateImportPlaylist() {
 }
 
+void BasePlaylistFeature::slotCreateImportPlaylistDirect(const QString& playlistFile) {
+    const QFileInfo fileInfo(playlistFile);
```
QFileInfo is a heavy class, caching disk access. Please try to create it only once.

daschuer commented on src/coreservices.cpp:
```diff
@@ -466,10 +483,17 @@ void CoreServices::initialize(QApplication* pApp) {
+            qWarning() << "Detected Playlist file!!";
+            const QString playlistFile = musicFiles.at(i);
+            qWarning() << playlistFile;
+            BasePlaylistFeature::slotCreateImportPlaylistDirect(playlistFile);
```
You need here an instance of BasePlaylistFeature. The slot prefix is used for Qt's signal slot connections, not here. 

I think you may create something like: 
```
void Library::importPlaylistFile(..) {
    m_pPlaylistFeature->importPlaylistFile(..);
}
```

and then you can call here: 
```
m_pLibrary->importPlaylistFile(playlistFile);
```

ronso0 replied:
Or just store the playlist fie path in `CmdlineArgs` and let `AutoDJFeature` constructor deal with it?

shivang0-0 replied:
> You need here an instance of BasePlaylistFeature. The slot prefix is used for Qt's signal slot connections, not here.
> 
> I think you may create something like:
> 
> ```
> void Library::importPlaylistFile(..) {
>     m_pPlaylistFeature->importPlaylistFile(..);
> }
> ```
> 
> and then you can call here:
> 
> ```
> m_pLibrary->importPlaylistFile(playlistFile);
> ```



> You need here an instance of BasePlaylistFeature. The slot prefix is used for Qt's signal slot connections, not here.
> 
> I think you may create something like:
> 
> ```
> void Library::importPlaylistFile(..) {
>     m_pPlaylistFeature->importPlaylistFile(..);
> }
> ```
> 
> and then you can call here:
> 
> ```
> m_pLibrary->importPlaylistFile(playlistFile);
> ```

in the above statement `m_pPlaylistFeature->importPlaylistFile(..);` gives
`pointer to incomplete class type "PlaylistFeature" is not allowed`.
Is there a work-around here @daschuer ?

ronso0 replied:
Tbh, you could figure this out on your own by searching the web for "pointer to incomplete class type not allowed".
You need to include the headers of classes you intend to use.

shivang0-0 replied:
Sorry, I got confused in the cyclic dependency of classes. I'll keep that in mind, for future instances. Thanks!!

ronso0 commented:
> Known issue: The [following calling](https://github.com/shivang0-0/mixxx/blob/playlist-import/src/coreservices.cpp#L496) is facing issue as BasePlaylistFeature is an abstract class.

...and CI tells you why, see https://github.com/mixxxdj/mixxx/actions/runs/3981086074/jobs/6824750974#step:9:641

ronso0 commented:
I'll repeat my proposal here in case it gets lost in the above review conversation:
How about simply storing the playlist file path in `CmdlineArgs` and fetch it in `AutoDJFeature` constructor?

shivang0-0 commented:
> I'll repeat my proposal here in case it gets lost in the above review conversation: How about simply storing the playlist file path in `CmdlineArgs` and fetch it in `AutoDJFeature` constructor?

Sorry, it was a busy Monday didn't have time to review the code, will look into that now. Thanks!!

shivang0-0 committed:
Update src/coreservices.cpp

initialise list in a single run

Co-authored-by: Daniel Schürmann <daschuer@mixxx.org>

shivang0-0 commented:
> I'll repeat my proposal here in case it gets lost in the above review conversation: How about simply storing the playlist file path in `CmdlineArgs` and fetch it in `AutoDJFeature` constructor?

Can you give a heads up where is the definition of AutoDJFeature

ronso0 commented:
I'll share some hints since I was curious if my proposal _could_ actually be implemented.
These hints are only useful if the sole purpose is to get all track locations from a playlist file and queue them into AutoDJ (i.e. don't import/create an actual playlist in the Playlists feature, don't create a temp playlist).
I think this would be much simpler (IMO) than doing it in coreservices. If you find this route actually works as I imagine ('works' like you can build it and performs as desired), I suggest you open a separate PR.

When searching the source code for `AutoDJFeature` you'd eventually end up here:
https://github.com/mixxxdj/mixxx/blob/09b2f614041e0627b8f1169f403b751817c77261/src/library/autodj/autodjfeature.cpp#L47

You can get the command line playlist file path without having a pointer to CommandLineArgs (like in coreservices) because you can get the instance anywhere via `CmdlineArgs::Instance()`, then get the music files from that.

Here's an example how to get a list of track locations from a playlistfile with static `Parser::parse`:
https://github.com/mixxxdj/mixxx/blob/09b2f614041e0627b8f1169f403b751817c77261/src/library/trackset/baseplaylistfeature.cpp#L475-L484

This is an example how track locations are handled when being dropped onto a library feature, regardless if they're already in the Mixxx library:
https://github.com/mixxxdj/mixxx/blob/09b2f614041e0627b8f1169f403b751817c77261/src/library/autodj/autodjfeature.cpp#L169-L181

The trackIds returned can be added to the AutoDJ playlist with https://github.com/mixxxdj/mixxx/blob/09b2f614041e0627b8f1169f403b751817c77261/src/library/dao/playlistdao.cpp#L1197

Have fun!

shivang0-0 committed:
implemeted playlist feature using void Library::importPlaylistFile()

This function tries to import playlist file when opened directly with file manager

shivang0-0 commented:
@daschuer, as per the requirement in code changes requested, I created a separate function aside from slot file, the build was successful, and function callings were verified without errors, but the imported playlist still isn't reflected at start-up. Can you review the code [here](https://github.com/shivang0-0/mixxx/blob/playlist-import/src/library/trackset/playlistfeature.cpp#L128-L149).

shivang0-0 commented:
> I'll share some hints since I was curious if my proposal _could_ actually be implemented. These hints are only useful if the sole purpose is to get all track locations from a playlist file and queue them into AutoDJ (i.e. don't import/create an actual playlist in the Playlists feature, don't create a temp playlist). I think this would be much simpler (IMO) than doing it in coreservices. If you find this route actually works as I imagine ('works' like you can build it and performs as desired), I suggest you open a separate PR.
> 
> When searching the source code for `AutoDJFeature` you'd eventually end up here:
> 
> https://github.com/mixxxdj/mixxx/blob/09b2f614041e0627b8f1169f403b751817c77261/src/library/autodj/autodjfeature.cpp#L47
> 
> You can get the command line playlist file path without having a pointer to CommandLineArgs (like in coreservices) because you can get the instance anywhere via `CmdlineArgs::Instance()`, then get the music files from that.
> 
> Here's an example how to get a list of track locations from a playlistfile with static `Parser::parse`:
> 
> https://github.com/mixxxdj/mixxx/blob/09b2f614041e0627b8f1169f403b751817c77261/src/library/trackset/baseplaylistfeature.cpp#L475-L484
> 
> This is an example how track locations are handled when being dropped onto a library feature, regardless if they're already in the Mixxx library:
> 
> https://github.com/mixxxdj/mixxx/blob/09b2f614041e0627b8f1169f403b751817c77261/src/library/autodj/autodjfeature.cpp#L169-L181
> 
> The trackIds returned can be added to the AutoDJ playlist with
> 
> https://github.com/mixxxdj/mixxx/blob/09b2f614041e0627b8f1169f403b751817c77261/src/library/dao/playlistdao.cpp#L1197
> 
> Have fun!

Looks promising!! Will try that for sure in a separate PR.
Thanks for the heads up!!

daschuer commented on src/coreservices.cpp:
```diff
@@ -466,10 +483,17 @@ void CoreServices::initialize(QApplication* pApp) {
         if (SoundSourceProxy::isFileNameSupported(musicFiles.at(i))) {
             m_pPlayerManager->slotLoadToDeck(musicFiles.at(i), i + 1);
+        } else if (isPlaylist(musicFiles.at(i))) {
+            qWarning() << "Detected Playlist file!!";
+            const QString playlistFile = musicFiles.at(i);
+            qWarning() << playlistFile;
```
Please combine these log outputs and make it a qDebug(). The text can also be improved. 
"Found playlist file ... iin the command line arguments" or such.

shivang0-0 replied:
formatted in latest commit

daschuer commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -536,6 +537,12 @@ void BasePlaylistFeature::slotCreateImportPlaylist() {
 }
 
+void Library::importPlaylistFile(const QString& playlistFile) {
+    qWarning() << "Reached basePlaylistfeature";
+    qWarning() << playlistFile;
```
can this be removed?

shivang0-0 replied:
removed

daschuer commented on src/library/trackset/baseplaylistfeature.cpp:
```diff
@@ -731,8 +738,8 @@ void BasePlaylistFeature::updateChildModel(int playlistId) {
-  * Clears the child model dynamically, but the invisible root item remains
-  */
+ * Clears the child model dynamically, but the invisible root item remains
+ */
```
We don't use Java style comments in new code.

shivang0-0 replied:
formatted in latest commit

daschuer commented on src/library/trackset/playlistfeature.cpp:
```diff
@@ -125,6 +125,29 @@ bool PlaylistFeature::dragMoveAcceptChild(const QModelIndex& index, const QUrl&
 }
 
+void PlaylistFeature::importPlaylistFile(const QString& playlistFile) {
+    qWarning() << "Reached playlistfeature";
+    qWarning() << playlistFile;
```
remove?

shivang0-0 replied:
removed

daschuer reviewed (changes requested):
The original bug reporter requests to "import" the playlist. 
This is what you have implemented. 

@ronso0 describes a different use case, to play a playlist in Auto DJ. This is what most audio players do when open it with a command line parameter. 

What is you plan to distinguish the cases?   

If you open Mixxx with the same playlist twice, it will be imported twice. Is this desired?

shivang0-0 commented:
Apologies for not being available for past 3 days. Had my system sent for a memory upgrade. Will be working on this now.

shivang0-0 commented:
> The original bug reporter requests to "import" the playlist. This is what you have implemented.
> 
> @ronso0 describes a different use case, to play a playlist in Auto DJ. This is what most audio players do when open it with a command line parameter.
> 
> What is you plan to distinguish the cases?
> 
> If you open Mixxx with the same playlist twice, it will be imported twice. Is this desired?

so far, the changes implemented shows no error but at the same time, the playlist is not available in the software's UI. Can you hint out where am I missing out.

shivang0-0 committed:
Formatting: fixed formatting and added Debug outputs

ronso0 commented on src/library/library.h:
```diff
@@ -95,6 +95,7 @@ class Library: public QObject {
     static const int kDefaultRowHeightPx;
 
     void setFont(const QFont& font);
+    void importPlaylistFile(const QString& playlistFile);
```
No implementation, no fun...
Did you forget to commit the implementation of this function?

ronso0 replied:
also, please check the failing CI jobs below

shivang0-0 replied:
> No implementation, no fun... Did you forget to commit the implementation of this function?

Oh!! got messed up while working on a separate PR for the issue. Will push the changes in time after checking the CI jobs.

ronso0 replied:
cool.
btw you're a volunteer contributor, no one is expecting you to deliver anything 'in time' or 'asap'
that just applies to core devs in case they messed something up.. :  )

KristjanESPERANTO commented:
The progress here looks promising! @shivang0-0, are you still working on this topic?

cr7pt0gr4ph7 commented:
I'm currently scanning the PR backlog to see if we could get the number of open PRs down. This PR can probably be closed as it has unresolved conflicts & is superseded/replaced by #16471.
